src/app/clusters/bindings/BindingManager.cpp

```diff
@@ -0,0 +1,200 @@
+            mPendingClusterMap.RemoveEntry(entry);
+            // Now retry
+            error = mAppServer->GetCASESessionManager()->FindOrEstablishSession(peer, &mOnConnectedCallback,
+                                                                                &mOnConnectionFailureCallback);
+        }
+    }
+    return error;
+}
+
+void BindingManager::HandleDeviceConnected(void * context, OperationalDeviceProxy * device)
+{
+    BindingManager * manager = static_cast<BindingManager *>(context);
+    manager->HandleDeviceConnected(device);
+}
+
+void BindingManager::HandleDeviceConnected(OperationalDeviceProxy * device)
+{
+    mPendingClusterMap.ForEachActiveObject([&](PendingClusterEntry * entry) -> Loop {
+        if (entry->GetPeerId() == device->GetPeerId())
+        {
+            SyncPendingClustersToPeer(device, entry);
+        }
+
+        return Loop::Continue;
+    });
+}
+
+void BindingManager::SyncPendingClustersToPeer(OperationalDeviceProxy * device, PendingClusterEntry * pendingClusters)
+{
+    for (const ClusterPath & path : *pendingClusters)
+    {
+        ClusterId cluster   = path.cluster;
+        EndpointId endpoint = path.endpoint;
+        for (uint8_t j = 0; j < EMBER_BINDING_TABLE_SIZE; j++)
+        {
+            EmberBindingTableEntry entry;
+            if (emberGetBinding(j, &entry) == EMBER_SUCCESS && entry.type != EMBER_UNUSED_BINDING && entry.clusterId == cluster &&
+                entry.local == endpoint && mBoundDeviceChangedHandler)
+            {
+                mBoundDeviceChangedHandler(entry.local, entry.remote, cluster, device, Optional<GroupId>());
+            }
+        }
+    }
+    mPendingClusterMap.RemoveEntry(pendingClusters);
+}
+
+void BindingManager::HandleDeviceConnectionFailure(void * context, PeerId peerId, CHIP_ERROR error)
+{
+    BindingManager * manager = static_cast<BindingManager *>(context);
+    manager->HandleDeviceConnectionFailure(peerId, error);
+}
+
+void BindingManager::HandleDeviceConnectionFailure(PeerId peerId, CHIP_ERROR error)
+{
+    // Simply release the entry, the connection will be re-established on need.
+    ChipLogError(AppServer, "Failed to establish connection to node 0x" ChipLogFormatX64, ChipLogValueX64(peerId.GetNodeId()));
+    mAppServer->GetCASESessionManager()->ReleaseSession(peerId);
+}
+
+CHIP_ERROR BindingManager::DisconnectDevice(FabricIndex fabric, NodeId node)
+{
+    VerifyOrReturnError(mAppServer != nullptr, CHIP_ERROR_INCORRECT_STATE);
+
+    FabricInfo * fabricInfo = mAppServer->GetFabricTable().FindFabricWithIndex(fabric);
+    VerifyOrReturnError(fabricInfo != nullptr, CHIP_ERROR_NOT_FOUND);
+    PeerId peer                 = fabricInfo->GetPeerIdForNode(node);
+    PendingClusterEntry * entry = mPendingClusterMap.FindEntry(peer);
+
+    VerifyOrReturnError(entry != nullptr, CHIP_ERROR_NOT_FOUND);
+
+    mAppServer->GetCASESessionManager()->ReleaseSession(peer);
```

Comment:
Once CASESessionManager grows LRU eviction, what does it mean to "leak the connection"?  I guess we could leak it in the sense that it still consumes resources and never gets evicted because we never create enough other connections?

I agree that handling this in a followup is OK.  Just need to make sure we do it.